### PR TITLE
[FLINK-15172][table-blink] Optimize the operator algorithm to lazily allocate memory

### PR DESCRIPTION
**Pull request by JingsongLi**
## What is the purpose of the change

Now after FLINK-14063 , operators will get all manage memory of TaskManager,  The cost of over allocate memory is very high, lead to performance regression of small batch sql jobs:

Allocate memory will have the cost of memory management algorithm.
Allocate memory will have the cost of memory initialization, will set all memory to zero. And this initialization will require the operating system to actually allocate physical memory.
Over allocate memory will squash the file cache too.
We can optimize the operator algorithm, apply lazy allocation, and avoid meaningless memory allocation.

## Brief change log

- Introduce LazyMemorySegmentPool
- Improve operators to lazily allocate memory

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): (no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no
  - The S3 file system connector: (no

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/sort/BinaryExternalSorter.java [hunk 1]
```diff
@@ -1054,7 +1018,7 @@ private void releaseEmptyBuffers() {
 					}
 				}
 			}
-			releaseCoreSegments();
+			sortReadMemory.forEach(LazyMemorySegmentPool::cleanCache);
```
Comment (TsReaper) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/sort/BinaryExternalSorter.java]: We don't need to manually clean cache, as memory pool will try to clean cache when a segment is returned.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/LazyMemorySegmentPool.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	public void close() {
+		if (this.pageUsage != 0) {
+			throw new RuntimeException(
+					"Should return all used memory before clean, page used: " + pageUsage);
+		}
+		cleanCache();
+	}
+
+	public void cleanCache() {
+		this.memoryManager.release(this.cachePages);
```
Comment (TsReaper) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/LazyMemorySegmentPool.java]: Shall we check if `this.cachePages.size() >= this.perRequestPages`?

Comment (JingsongLi) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/LazyMemorySegmentPool.java]: No, this is a manual release of excess memory. This frequency will not be too high.

**Comment by JingsongLi** [comment]
travis passed in https://travis-ci.org/JingsongLi/flink/builds/634138477

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/LazyMemorySegmentPool.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+	@Override
+	public void returnAll(List<MemorySegment> memory) {
+		this.pageUsage -= memory.size();
+		if (this.pageUsage < 0) {
+			throw new RuntimeException("Return too more memories.");
+		}
+		this.cachePages.addAll(memory);
+
+		if (this.freeMore && this.cachePages.size() >= this.perRequestPages) {
```
Comment (TsReaper) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/LazyMemorySegmentPool.java]: Is this `freeMore` mode actually in use? How much does it impact the performance?

Comment (JingsongLi) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/LazyMemorySegmentPool.java]: not used now, I'll delete it.

**Comment by JingsongLi** [comment]
Thanks @TsReaper for the reviewing. Merged.

**Pull request closed without merging**